_posts/2020-09-29-introducing-torch-for-r/introducing_torch_for_R.Rmd

```diff
@@ -0,0 +1,424 @@
+
+Here are the first 32 characters:
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Kuzushiji MNIST."}
+knitr::include_graphics("images/kmnist.png")
+```
+
+#### Dataset
+
+The following code will download the data, separately for training and test sets.
+
+```{r}
+dir <- "."
+
+train_ds <- kmnist_dataset(
+  dir,
+  download = TRUE,
+  train = TRUE,
+  transform = function(x) {
+    x <- x$to(dtype = torch_float())/256
+    x[newaxis,..]
+  }
+)
+
+test_ds <- kmnist_dataset(
+  dir,
+  download = TRUE,
+  train = FALSE,
+  transform = function(x) {
+    x <- x$to(dtype = torch_float())/256
+    x[newaxis,..]
+  }
+)
+
+```
+
+Note the `transform` argument. The passed-in function applies two transformations: First, it normalizes the pixels to the range between 0 and 1. Then, it adds another dimension in front. Why?
+
+Contrary to what you might expect -- if until now, you've been using `keras` -- the additional dimension is *not* the batch dimension. Batching will be taken care of by the `dataloader`, to be introduced next. Instead, this is the *channels* dimension that in `torch`, is found *before* the width and height dimensions by default.
+
+One thing I've found to be extremely useful about `torch` is how easy it is to inspect objects. Even though we're dealing with a `dataset`, a custom object, and not an R array or even a `torch` tensor, we can easily peek at what's inside. Indexing in `torch` is 1-based, conforming to the R user's intuitions. Consequently,
+
+```{r}
+train_ds[1]
+```
+
+gives us the first element in the dataset, an R *list* of two tensors corresponding to input and target, respectively. (We don't reproduce the output here, but you can see for yourself in the notebook.)
+
+Let's inspect the shape of the input tensor:
+
+```{r}
+train_ds[1][[1]]$size()
+```
+
+    [1]  1 28 28
+
+Now we have the data, we need someone to feed them to a deep learning model, nicely batched and all. In `torch`, this is the task of data loaders.
+
+#### Data loader
+
+Each of training and test sets get their own data loader:
+
+```{r}
+train_dl <- dataloader(train_ds, batch_size = 32, shuffle = TRUE)
+test_dl <- dataloader(test_ds, batch_size = 32)
+```
+
+Again, `torch` makes it easy to verify we did the correct thing. To take a look at the content of the first batch, do
+
+```{r}
+train_iter <- train_dl$.iter()
+train_iter$.next()
+```
+
+Functionality like this may not seem indispensable when working with a well-known dataset, but will turn out to be very useful when a lot of domain-specific pre-processing is required.
+
+Now that we've seen how to load data, all prerequisites are fulfilled for visualizing them. Here is the code that was used to display the first batch of characters, above:
+
+```{r, fig.asp=0.5, fig.width=8}
+par(mfrow = c(4,8), mar = rep(0, 4))
+images <- train_dl$.iter()$.next()[[1]][1:32, 1, , ] 
+images %>%
+  purrr::array_tree(1) %>%
+  purrr::map(as.raster) %>%
+  purrr::iwalk(~{plot(.x)})
```

Comment:
Looks like the image is missing in the preview

Reply:
it shouldn't be there - that's why I say
> Here is the code that was used to display the first batch of characters, above: